File: apps/image-api/src/services/cache.ts
import Redis from "ioredis";
import { config } from "../config";

let redis: Redis | null = null;

export function getRedis(): Redis | null {
  if (redis) return redis;
  try {
    redis = new Redis(config.redisUrl, {
      maxRetriesPerRequest: 1,
      lazyConnect: true,
    });
    return redis;
  } catch {
    return null;
  }
}

export async function cacheGet<T>(key: string): Promise<T | null> {
  const client = getRedis();
  if (!client) return null;
  try {
    if (client.status !== "ready") await client.connect();
    const val = await client.get(key);
    return val ? (JSON.parse(val) as T) : null;
  } catch {
    return null;
  }
}

export async function cacheSet(
  key: string,
  value: unknown,
  ttlSeconds = 60,
): Promise<void> {
  const client = getRedis();
  if (!client) return;
  try {
    if (client.status !== "ready") await client.connect();
    await client.set(key, JSON.stringify(value), "EX", ttlSeconds);
  } catch {
    // cache is best-effort
  }
}

export async function cacheInvalidate(pattern: string): Promise<void> {
  const client = getRedis();
  if (!client) return;
  try {
    if (client.status !== "ready") await client.connect();
    const keys = await client.keys(pattern);
    if (keys.length) await client.del(...keys);
  } catch {
    // ignore
  }
}

export async function pingRedis(): Promise<boolean> {
  const client = getRedis();
  if (!client) return false;
  try {
    if (client.status !== "ready") await client.connect();
    return (await client.ping()) === "PONG";
  } catch {
    return false;
  }
}

File: app/services/cache.py
"""Best-effort Redis cache, mirroring services/cache.ts."""

import json
import logging
from typing import Any

import redis

from config import config

logger = logging.getLogger(__name__)

_client: redis.Redis | None = None


def get_redis() -> redis.Redis | None:
    global _client
    if _client is not None:
        return _client
    try:
        _client = redis.Redis.from_url(
            config.REDIS_URL,
            decode_responses=True,
            socket_connect_timeout=1,
            socket_timeout=1,
        )
        return _client
    except Exception:
        return None


def cache_get(key: str) -> Any | None:
    client = get_redis()
    if client is None:
        return None
    try:
        value = client.get(key)
        return json.loads(value) if value else None
    except Exception:
        return None


def cache_set(key: str, value: Any, ttl_seconds: int = 60) -> None:
    client = get_redis()
    if client is None:
        return
    try:
        client.set(key, json.dumps(value), ex=ttl_seconds)
    except Exception:
        pass  # cache is best-effort


def cache_invalidate(pattern: str) -> None:
    client = get_redis()
    if client is None:
        return
    try:
        keys = client.keys(pattern)
        if keys:
            client.delete(*keys)
    except Exception:
        pass


def ping_redis() -> bool:
    client = get_redis()
    if client is None:
        return False
    try:
        return client.ping()
    except Exception:
        return False
